test: Add test on skip heights in CBlockIndex

optout21 opened the pull request:
The skip height values computed by the (internal) function `GetSkipHeight()`, and the `CBlockIndex::BuildSkip()` method are not tested directly, and the skip logic is not well documented. To improve test coverage, a new test is added, to verify `CBlockIndex::BuildSkip()` and the skip height bit-manipulation logic.

Note: the original version contained a test for the complexity of the `GetAncestor()` algorithm, whose performance is greatly determined by the skip height logic. This test was out-scoped. (see: https://github.com/bitcoin/bitcoin/pull/33661#issuecomment-4486353485)

The motivation is to document the skip value computation through a test. (The issued was noticed while reviewing #33515.)

optout21 commented:
> For reference https://maflcko.github.io/b-c-cov/test_bitcoin.coverage/src/chain.cpp.gcov.html is pretty well covered otherwise, but these tests can help with the specifics.

Thanks for the link! Indirectly it is tested extensively, as the skip mechanism is used while searching in the chain. However, there are no tests focusing on this specifically, and the why and how on the exact skip generation algorithm is not documented.

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+{
+    // The test collects and analyses the distribution of the
+    // skip difference numbers, and checks some properties:
+    // non-uniform distribution with most values small but
```
it seems to test more than non-uniform.

optout21 replied:
I check the avg, which is computationally easier to obtain. It would make sense to check the median as well, and maybe compare the median with the average. However, I prefer to avoid the O(N log N) complexity of the sorting.

optout21 replied:
Added median check as well. Much lower than the avg!

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    int max_diff = 0;
+    int total_diff = 0;
+    for(int i = 1; i < n; ++i) {
+        BOOST_CHECK(block_index[i]->nHeight == i);
```
this sounds like a `BOOST_REQUIRE_EQUAL` instead

optout21 replied:
The check on `nHeight` has been removed, as it was essentially just validating the setup procedure, not the actual logic under test. This comment is thus outdated, resolving.

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    int count_diff_smaller_16 = 0;
+    int max_diff = 0;
+    int total_diff = 0;
+    for(int i = 1; i < n; ++i) {
```
how come we're iterating one less than before?

l0rinc reviewed (changes requested):
Concept ACK, during review I also understood this behavior better :)
I have a few suggestions for simplification and am throwing you into the deep end by asking for a property based (fuzz) test as well to check certain invariants. Not a must-have, but seems simple enough. I also suggest simplifying the distribution test to just check for the median instead.

optout21 commented:
Review comments addressed, mostly minor:

- Reduce the number of iterations to 10000 (from 30000), for faster test execution.
- Add comments on choice of test inputs.
- Use BOOST_REQUIRE_EQUAL or BOOST_CHECK_GT where possible.

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    // Create genesis block
+    auto genesis = std::make_unique<CBlockIndex>();
+    genesis->nHeight = 0;
+    block_index.push_back(std::move(genesis));
```
I see you've moved it as well, clang complains with:
> Clangd: Moving a temporary object prevents copy elision

optout21 commented:
Changes applies:
- Fix lint issue
- Added check on the median as well
- some minor formatting

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
 
+BOOST_AUTO_TEST_CASE(get_skip_height_test)
+{
+    BOOST_CHECK(GetSkipHeight(0b0001011101101000) == 0b0001011101100000);
```
I did that for the longer values, for shorter values with only a few ones is not really need.

l0rinc replied:
Can we do it for the shorter ones as well? it's easier to compare them and more consistent with the rest

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        block_index.push_back(std::move(new_index));
+    }
+
+    // Analyze the diff (in height) of each element and its pskip
```
Added explanatory comments

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,78 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    block_index.push_back(std::move(genesis));
+
+    // Build a chain
+    const auto n{30000};
```
Done

optout21 commented on src/chain.cpp:
A fuzz test for this functionality (essentially a one-liner) seems a bit too much.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,113 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+                                    0b0110101101010000);
+    BOOST_CHECK_EQUAL(GetSkipHeight(0b0110101101011001),
+                                    0b0110101101000001);
+    // All values 0--10
```
"0--10" meant "from 0 to 10 inclusive". Changed to "0-10" to avoid parsing issues.

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +45,110 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    for(int i{1}; i < n; ++i) {
+        auto new_index = std::make_unique<CBlockIndex>();
+        new_index->pprev = block_index.back().get();
+        BOOST_CHECK(new_index->pprev);
```
nit: should likely be a `BOOST_REQUIRE` since this is still the setup phase, not yet the validation stage

optout21 replied:
True; check was removed (with the change to on-stack allocation).

l0rinc reviewed (approved):
ACK ea998e18f6d64f485d4e2d85a1028474fb35120a

Thanks for taking my suggestion, the code is very clean and easy to understand now.
I left some nits I still saw, but they're not blocking.

optout21 commented:
Minor nits applied

<details>
<summary>git diff ea998e18f6d64f485d4e2d85a1028474fb35120a 02df44465218a49cc488fd50bf05ab2fd203c69c
</summary>

```diff
@@ -114,7 +114,7 @@ BOOST_AUTO_TEST_CASE(skip_height_properties_test)
     for(int i{1}; i < n; ++i) {
         auto new_index = std::make_unique<CBlockIndex>();
         new_index->pprev = block_index.back().get();
-        BOOST_CHECK(new_index->pprev);
+        BOOST_REQUIRE(new_index->pprev);
         new_index->nHeight = new_index->pprev->nHeight + 1;
         new_index->BuildSkip();
         block_index.push_back(std::move(new_index));
@@ -131,16 +131,12 @@ BOOST_AUTO_TEST_CASE(skip_height_properties_test)
     }
 
     // Analyze the diffs
-    // Compute the average
-    const auto avg = double(std::accumulate(skip_diffs.begin(), skip_diffs.end(), 0)) / double(skip_diffs.size());
-    // Compute the median
+    // Compute the average, median, maximum
+    const auto avg{double(std::accumulate(skip_diffs.begin(), skip_diffs.end(), 0)) / double(skip_diffs.size())};
     std::ranges::sort(skip_diffs);
     const auto median{skip_diffs[n / 2]};
-    // Take the maximum
     const auto max{skip_diffs[skip_diffs.size() - 1]};
 
-    // Assert some parameters of the distribution
-
     // The median skip diff is lower than the average (non-uniform distribution)
     BOOST_CHECK_LT(median, avg);
     // There are some large skip diffs (more than half of n):
```
</details>

optout21 commented:
Very minor touch: moving of a curly nit-fix, rebase (noop); mostly to get CI to run again (had some apparently temporary issue).

(e2e25f4947e65869b014483be70b764a62bd0df7 -> 1a0423cf25bb60f98a29e06c39b7db2998860cd9)

l0rinc commented:
reACK 1a0423cf25bb60f98a29e06c39b7db2998860cd9

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +41,80 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        auto heightWalk = start_height;
+        auto iteration_count{0};
+        while (heightWalk > target_height) {
+            int heightSkip = GetSkipHeight(heightWalk);
```
both commits need to pass, we should migrate the

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        // Asserts on the number of iterations:
+        // always less than 100 and less than 5% of the distance
+        BOOST_CHECK_LT(iteration_count, 100);
+        BOOST_REQUIRE(delta > 0);
```
```suggestion
        BOOST_REQUIRE_GT(delta, 0);
```

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    FastRandomContext ctx(/*fDeterministic =*/true);
+    for (auto i{0}; i < runs; ++i) {
+        // pick a height in the second half of the chain
+        auto start_height = chain_size / 2 + ctx.randrange(chain_size / 2);
```
To make sure we don't have signed/unsigned problems here (since we're subtracting from this later), we might as well make it explicitly `int`:

```suggestion
        int start_height{chain_size / 2 + ctx.randrange(chain_size / 2)};
```

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        // pick a height in the second half of the chain
+        auto start_height = chain_size / 2 + ctx.randrange(chain_size / 2);
+        BOOST_REQUIRE(start_height > 0 && start_height < chain_size);
+        auto& start_p = block_index[start_height];
```
can you use brace init consistently to avoid narrowing?

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+
+    const auto runs{1000};
+    const auto min_distance{500};
+    FastRandomContext ctx(/*fDeterministic =*/true);
```
making it deterministic kinda' defeats the purpose in my opinion (it's a low bar to construct something that accidentally works, a lot more difficult to find the exact properties of a system that work less predictable values)

nit: extra space before `=`

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        BOOST_CHECK_LT(iteration_count, 100);
+        BOOST_REQUIRE(delta > 0);
+        auto rel_iter_count = double(iteration_count) / double(delta);
+        BOOST_CHECK_LT(rel_iter_count, 0.05);
```
can we avoid floating point arithmetic here?

```suggestion
        BOOST_CHECK_LT(iteration_count * 20, delta);
```

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+                (heightSkip == target_height ||
+                (heightSkip > target_height && !(heightSkipPrev < heightSkip - 2 &&
+                                        heightSkipPrev >= target_height)))) {
+                // Only follow pskip if pprev->pskip isn't better than pskip->pprev.
```
Instead of commenting, can we extract the logic to variables that explain what happens here?
After a lot of fiddling with this part, it seems to me this may explain it a bit better:
```C++
// Perform traversal from start to target, skipping, as implemented in `GetAncestor`
const auto* pindex_walk{block_index[start_height].get()};
int iteration_count{0};
for (int height_walk{start_height}; height_walk > target_height; ++iteration_count) {
    BOOST_REQUIRE(pindex_walk && pindex_walk->pprev);
    const auto* prev{pindex_walk->pprev};
    const auto* skip{pindex_walk->pskip};

    const int height_skip{skip ? skip->nHeight : 0};
    const int height_skip_prev{prev->pskip ? prev->pskip->nHeight : 0};
    if (skip && height_skip >= target_height && std::max(height_skip_prev, target_height) >= height_skip - 2) {
        saw_skip |= true;
        pindex_walk = skip;
        height_walk = height_skip;
    } else {
        saw_prev |= true;
        pindex_walk = prev;
        --height_walk;
    }
}
```

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+                                        heightSkipPrev >= target_height)))) {
+                // Only follow pskip if pprev->pskip isn't better than pskip->pprev.
+                pindexWalk = pindexWalk->pskip;
+                heightWalk = heightSkip;
```
we should assert that this will always converge towards genesis (i.e. it's decreasing):
```C++
                BOOST_REQUIRE(skip->nHeight < height_walk);
```

optout21 replied:
Added now.

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    BOOST_CHECK_EQUAL(GetSkipHeight(10), 8);
+}
+
+BOOST_AUTO_TEST_CASE(skip_height_analysis)
```
Seems fast enough:
> ./build/bin/test_bitcoin --run_test=chain_tests/skip_height_analysis  0.53s user 0.02s system 32% cpu 1.714 total

l0rinc reviewed (changes requested):
I might have gone a bit overboard, but I have tried explaining every change that I suggested in context.

For convenience, here's the full local diff I did during review (this is the latest state, some of the suggestions may reflect a previous one):
```C++
BOOST_AUTO_TEST_CASE(get_skip_height_test)
{
    // Even heights: clear the lowest set bit.
    BOOST_CHECK_EQUAL(GetSkipHeight(0b00010010),
                                    0b00010000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b00100010),
                                    0b00100000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b01000010),
                                    0b01000000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b00010100),
                                    0b00010000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b00011000),
                                    0b00010000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10101010),
                                    0b10101000);
    // Odd heights: clear the 2nd and 3rd set bits.
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10010011),
                                    0b10000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10100011),
                                    0b10000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b11000011),
                                    0b10000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10010101),
                                    0b10000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10011001),
                                    0b10000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b10101011),
                                    0b10100001);
    // Some longer random values (even and odd)
    BOOST_CHECK_EQUAL(GetSkipHeight(0b0001011101101000),
                                    0b0001011101100000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b0001011101101001),
                                    0b0001011101000001);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b0110101101011000),
                                    0b0110101101010000);
    BOOST_CHECK_EQUAL(GetSkipHeight(0b0110101101011001),
                                    0b0110101101000001);
    // All values 0-10
    BOOST_CHECK_EQUAL(GetSkipHeight(0), 0);
    BOOST_CHECK_EQUAL(GetSkipHeight(1), 0);
    BOOST_CHECK_EQUAL(GetSkipHeight(2), 0);
    BOOST_CHECK_EQUAL(GetSkipHeight(3), 1);
    BOOST_CHECK_EQUAL(GetSkipHeight(4), 0);
    BOOST_CHECK_EQUAL(GetSkipHeight(5), 1);
    BOOST_CHECK_EQUAL(GetSkipHeight(6), 4);
    BOOST_CHECK_EQUAL(GetSkipHeight(7), 1);
    BOOST_CHECK_EQUAL(GetSkipHeight(8), 0);
    BOOST_CHECK_EQUAL(GetSkipHeight(9), 1);
    BOOST_CHECK_EQUAL(GetSkipHeight(10), 8);
}

BOOST_AUTO_TEST_CASE(skip_height_analysis)
{
    // This test case indirectly tests the `GetSkipHeight` function,
    // and asserts that its properties are suitable for performant
    // ancestor search.
    // It does so by imitating the `GetAncestor` function to
    // look for an ancestor, and ensuring that an arbitrary
    // ancestor is found in no more steps than linear traversal.

    constexpr int chain_size{200'000};
    std::vector<std::unique_ptr<CBlockIndex>> block_index;
    block_index.reserve(chain_size);
    for (int i{0}; i < chain_size; ++i) {
        auto index{std::make_unique<CBlockIndex>()};
        index->pprev = i ? block_index.back().get() : nullptr;
        index->nHeight = i;
        index->BuildSkip();
        block_index.push_back(std::move(index));
    }

    bool saw_skip{false}, saw_prev{false};

    FastRandomContext rng(/*fDeterministic=*/false);
    for (int run{0}; run < 1'000; ++run) {
        const int start_height{(chain_size - 1) - rng.randrange(chain_size / 2)};   // pick a height in the second half of the chain
        constexpr int min_distance{100};
        const int delta{min_distance + rng.randrange(start_height - min_distance)}; // pick a target height, earlier by at least min_distance
        const int target_height{start_height - delta};

        BOOST_REQUIRE(start_height > 0 && start_height < chain_size);
        BOOST_REQUIRE(target_height > 0 && target_height < chain_size && target_height < start_height);
        BOOST_REQUIRE(delta > 0);

        // Perform traversal from start to target, skipping, as implemented in `GetAncestor`
        const auto* pindex_walk{block_index[start_height].get()};
        int iteration_count{0};
        for (int height_walk{start_height}; height_walk > target_height; ++iteration_count) {
            BOOST_REQUIRE(pindex_walk);
            const auto* const prev{Assert(pindex_walk->pprev)};
            const auto* const skip{Assert(pindex_walk->pskip)};
            BOOST_REQUIRE(prev->pskip);

            const bool prev_skip_better{prev->pskip->nHeight >= target_height && prev->pskip->nHeight < skip->nHeight - 2};
            if (skip->nHeight == target_height || (skip->nHeight > target_height && !prev_skip_better)) {
                saw_skip = true;
                pindex_walk = skip;
                BOOST_REQUIRE(skip->nHeight < height_walk);
                height_walk = skip->nHeight;
            } else {
                saw_prev = true;
                pindex_walk = prev;
                --height_walk;
            }
            BOOST_REQUIRE(height_walk >= target_height && height_walk == pindex_walk->nHeight);
        }

        // End-of-walk invariants.
        BOOST_CHECK_EQUAL(pindex_walk->nHeight, target_height);
        BOOST_CHECK_LE(iteration_count, 110); // See GetSkipHeight

        // Cross-check with the real implementation.
        const auto* ancestor{block_index[start_height]->GetAncestor(target_height)};
        BOOST_CHECK_EQUAL(ancestor, block_index[target_height].get());
    }

    BOOST_CHECK(saw_skip && saw_prev);
}
```

optout21 commented:
> 2 errors generated.
> ```

Fixed in 464ce57745894d4038c970c5baac3c9da5933c19

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        BOOST_REQUIRE(target_height > 0 && target_height < chain_size && target_height < start_height);
+
+        // Perform traversal from start to target, skipping, as implemented in `GetAncestor`
+        const CBlockIndex* pindexWalk = start_p.get();
```
It is saved in a local variable because it is also used at the end of the test method (`auto p_ancestor{start_p->GetAncestor(target_height)};`)

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        auto iteration_count{0};
+        while (heightWalk > target_height) {
+            int heightSkip = GetSkipHeight(heightWalk);
+            int heightSkipPrev = GetSkipHeight(heightWalk - 1);
```
Same here, I'd rather keep it to match the original code.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+        auto new_index{std::make_unique<CBlockIndex>()};
+        new_index->pprev = block_index.back().get();
+        BOOST_CHECK(new_index->pprev);
+        new_index->nHeight = new_index->pprev->nHeight + 1;
```
Makes sense, but I keep it as it is. This way it shows better, the for each object `nHeight = pprev->nHieght + 1` holds. Plus this is done the same way in the original `chain_test` test.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    std::vector<std::unique_ptr<CBlockIndex>> block_index;
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
```
Makes sense, but it make sense to keep the genesis special in this test.

optout21 commented:
> So I think it's worth considering one of those two approaches

I agree, point taken. I have considered adding the counting to the product code, but it's not a good idea to modify/extend the production code for testing purpose only, even just slightly. You are right (both of you) that duplicated likely will drift away in the futures, so there is no point in keeping it verbatim copy. I was just thinking that it's easier to find the original code if it's a verbatim copy, but that's a minor consideration.
So I give up on the verbatim copy, and I can do @l0rinc 's suggestions (code style).

optout21 commented:
Regarding model/approximation (https://github.com/bitcoin/bitcoin/pull/33661#discussion_r2620420093):

- If there was only the 1-bit clearing (applied for even numbers), the expected average for the skip is $log(y) - 1$. For the number of iterations, the approximation is be $(y-x)/log(y)$.
- The two-bit clearing is a more complex, but it approximates to $(log(y) - 1)(log(y) - 1)/2$. Combining the two (for 50% even numbers 1-bit, for 50% odd numbers 2-bit) yields $(log(y) - 1)(log(y) + 1)/2$. For iteration count estimate: $2*(y-x)/(log(y) - 1)(log(y) + 1)$ (having the order of $(y-x)/log^2(y)$ ).
- However, we are interested of the sum of skips in a particular traversal, not random. There is the effect that as bits are cleared, they tend to progress towards numbers with lesser low-order bits set, which mean higher skips. Modeling this is beyond me (and not warranted by this simple test)

- I looked empirically at the number of iterations vs. the log of the distance ($y-x$). Apparently, the function $8*log(y-x)-40$ gives a good-enough fit for the upper bound (visually derived; log is 2-based). See chart:

<img width="1106" height="800" alt="pr33661_chart1" src="https://github.com/user-attachments/assets/9654699b-9159-49df-a25b-33f6287feabe" />

I will use this formula as an upper bound in the test (together with the fixed overall-max of `115`, also empirically measured).

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    for (auto i{0}; i < runs; ++i) {
+        // pick a height in the second half of the chain
+        auto start_height = chain_size / 2 + ctx.randrange(chain_size / 2);
+        BOOST_REQUIRE(start_height > 0 && start_height < chain_size);
```
Done; delta check no longer needed.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    }
+
+    const auto runs{1000};
+    const auto min_distance{500};
```
Details in https://github.com/bitcoin/bitcoin/pull/33661#issuecomment-3659695776

optout21 replied:
> This will make boundary conditions less likely, running it with `constexpr int min_distance{100};` revealed new failures for the randomized case for me locally.

The "5%" rule was very rough, and broke with short distances, you are right. I added it as the global maximum check does not enforce much for short distances. But now the formula-based boundary is stricter over a wide range of distance values.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -108,11 +109,11 @@ BOOST_AUTO_TEST_CASE(skip_height_analysis)
-        BOOST_REQUIRE_GT(delta, 0);
-        const auto rel_iter_count{double(iteration_count) / double(delta)};
-        BOOST_CHECK_LT(rel_iter_count, 0.05);
+        // Absolute value maximum (derived empirically)
```
You're correct, the first 3 commits should be merged, they just reflect the evolution over the review, and there's no need for that. I will do that.

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -108,11 +109,11 @@ BOOST_AUTO_TEST_CASE(skip_height_analysis)
-        BOOST_REQUIRE_GT(delta, 0);
-        const auto rel_iter_count{double(iteration_count) / double(delta)};
-        BOOST_CHECK_LT(rel_iter_count, 0.05);
+        // Absolute value maximum (derived empirically)
```
Indeed. I adjust the parameter to work with huge number of iterations as well (the formula is empirically derived, as the note says). Thanks for trying out!
BTW, the test is deterministic, so sporadic failing is not an issue (if it passes with a given number of iteration, it does so consistently).

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +42,139 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    // metric than CPU time.
+
+    // Build a chain
+    const int chain_size{1'000'000};
```
That's a lot of allocations, isn't this problematic with our CI sanitizers?

l0rinc reviewed (changes requested):
I think we can get rid of a few unique pointers to speed up and simplify the test.

The PR description needs to be updated now since the added test doesn’t test a distribution anymore, the low-level test is commit 3, and the non-test change (making GetSkipHeight linkable) is commit 1.
The last commit still claims: `The method is changed to be not static and inline, so that the test can link with it.` which is done in a different commit now.

And `are not tested and not well documented` is bit overstated since there is existing skiplist coverage (src/test/skiplist_tests.cpp) that exercises BuildSkip() + GetAncestor(). I think we should move the new changes there.

Lastly, a few of my comments were resolved but not addressed, which is a bit frustrating, could you please recheck them and either explain why my assessment was incorrect or apply it to the change?

optout21 commented:
> I think we can get rid of a few unique pointers to speed up and simplify the test.
> [...]
> And `are not tested and not well documented` is bit overstated since there is existing skiplist coverage (src/test/skiplist_tests.cpp) that exercises BuildSkip() + GetAncestor(). I think we should move the new changes there.
> [...]
> Lastly, a few of my comments were resolved but not addressed, which is a bit frustrating, could you please recheck them and either explain why my assessment was incorrect or apply it to the change?

Thx. I will address the separate review comments.
The allocation remark is very useful.
Thx for mentioning `skiplist_tests.cpp`, it was off my radar (I initially searched on `GetSkipHeight` in tests).
I have re-checked Resolved comments (~42 of them), and indeed, I found 4 where the closing reason was not very clear. Sorry for that. I unresolved them and I will address or explain each of them.

optout21 commented:
WIP status: suggestions addressed, TODO:
- fixup commits
- review/update commit messages and PR description
- move tests from `chain_tests.cpp` to `skiplist_tests.cpp`

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +42,139 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    // ancestor is found in a number of steps that is much lower
+    // than simple linear traversal.
+    // This could be tested by a performance test as well, but here the
+    // number of iterations can be measured, that gives a more precice
```
Done

optout21 replied:
Fixed

optout21 commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +42,139 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
+    for(auto i{1}; i < chain_size; ++i) {
```
Done (91fce2c893cfe88180feae4d22c102813352548f)

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -46,6 +46,7 @@ BOOST_AUTO_TEST_CASE(get_skip_height_test)
 {
     // Even values: the rightmost set bit is zeroed
     // Test with various even values with at least 2 bits set
+    // clang-format off
```
we can do that, but it's obvious that this done is on purpose and we don't have a format linter that we have to satisfy

l0rinc commented on src/test/chain_tests.cpp:
```diff
@@ -41,6 +42,139 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
+    for(auto i{1}; i < chain_size; ++i) {
```
Thanks, though it's usually easier to review if it's pushed atomically (all changes addressed), otherwise we're just getting notifications and there's no way to act on it yet.

Also note that most reviewers prefer having a clear commit history that doesn't reflect its evolution inside the same PR - instead of adding new commits on top to modify code that we've just introduced, please do an interactive rebase on the existing commits.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +189,142 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+    // Repeat iterations. The test should work with much higher numbers as well,
+    // kept reasonable to keep execution time low
+    for (auto i{0}; i < 1'000; ++i) {
+        // pick a height in the second half of the chain
```
seems this was reverted in-between rebases

l0rinc reviewed (approved):
I left a few nits that I would prefer if we still addressed, but the current version is quite clean.

Some changes were lost during the rebase, if you could go over the comments again and make sure they are part of the final state, I'd appreciate it. I tried doing the same, left some notes.

ACK 8dc3e3ede97e3deea277f1d88583cc4be2a2b208

optout21 commented:
Thanks for the thorough checking! With adjusted boundary check I'm confident now that this holds for ALL possible subchains (with the given parameters: chain length of `1'000'000` and minimum length of `100`).
If you prefer, I'm open to reduce the test to a much simpler and relaxed check (e.g. `"< 200"`), but I think the current version is more informative.

Change:
- Boundary check values have been adjusted, to be more resilient for all possible subchains, not just the ones used in the test. The new formula is `8 * std::bit_width<unsigned>(delta) - 29` (a step-like function).

Additional info:

- A more thorough measurement was done. From the charts below it is visible that the formula used was not sufficient, it was not an upper bound in all cases (esp. for longer subchain).
- The chart confirms that the max iteration count has strong correlation with the delta.
- The absolute upper limit of `130` still holds
- The test passed with a test iteration count of `2'000'000'000`, instead of the `1000` used in the test.
- I reproduce some charts, plotting the measured maximum iteration count against the length of the subchain ('delta'). Several percentiles are also plotted for the distribution of itercounts for a given delta: 10th, 50th, 90th. The second chart is log-scale x axis, the other two are closeups. The data is not exhaustive.
- I also tried the test with non-deterministic RNG, ran 5000 times successfully.
- The `bit_width` used is a nice shortcut for `ceil(log(x)/log(2))`, without using floating point arithmetics (thanks for the hint, @l0rinc )

<details>
<summary>git diff 45a5c5f3d63da22aca7f14b9eb0203ffa827a1a7 527b186bb7908bdd19da7a8277727dd1079f7c95</summary>

```diff
diff --git a/src/test/skiplist_tests.cpp b/src/test/skiplist_tests.cpp
index d7d18cf76e..56a15d5203 100644
--- a/src/test/skiplist_tests.cpp
+++ b/src/test/skiplist_tests.cpp
@@ -6,6 +6,7 @@
 #include <test/util/setup_common.h>
 
+#include <bit>
 #include <cmath>
 #include <vector>
@@ -317,8 +318,8 @@ BOOST_AUTO_TEST_CASE(skip_height_analysis)
         BOOST_CHECK_LE(iteration_count, 130);
         // Dynamic bound, function of distance. Formula derived empirically.
-        const auto bound_log_distance{std::ceil(8.3 * std::log(delta) / std::log(2) - 35)};
-        BOOST_CHECK_LE(double(iteration_count), bound_log_distance);
+        const auto bound_log_distance{8 * std::bit_width<unsigned>(delta) - 29};
+        BOOST_CHECK_LE(iteration_count, bound_log_distance);
 
         // Also check that `GetAncestor` gives the same result
```
</details>

<img width="2200" height="1397" alt="charts4" src="https://github.com/user-attachments/assets/518594ce-55a4-4534-ac01-5bc9be97e72b" />

sipa commented:
I wrote an exhaustive search to find all the worst possible cases:

```c++
#include <stdint.h>
#include <vector>
#include <iostream>

namespace {

int constexpr inline InvertLowestOne(int n) { return n & (n - 1); }
int constexpr static inline GetSkipHeight(int height) {
    if (height < 2) return 0;
    return (height & 1) ? InvertLowestOne(InvertLowestOne(height - 1)) + 1 : InvertLowestOne(height);
}

void Analyse(int target_height, std::vector<std::pair<int, int>>& max_by_delta) noexcept {
    std::vector<int> steps_by_delta(1000000 - target_height);
    steps_by_delta[0] = 0;
    for (int delta = 1; delta <= 999999 - target_height; ++delta) {
        int starting_height = target_height + delta;
        int walk = starting_height;
        int skip = GetSkipHeight(walk);
        int skipprev = GetSkipHeight(walk - 1);
        if (skip == target_height || (skip > target_height && !(skipprev < skip - 2 && skipprev >= target_height))) {
            steps_by_delta[delta] = steps_by_delta[skip - target_height] + 1;
        } else {
            steps_by_delta[delta] = steps_by_delta[delta - 1] + 1;
        }
        if (steps_by_delta[delta] > max_by_delta[delta].first) {
            max_by_delta[delta].first = steps_by_delta[delta];
            max_by_delta[delta].second = target_height;
            std::cout << std::unitbuf << "delta=" << delta << ": " << starting_height << "->" << target_height << ": " << steps_by_delta[delta] << " steps\n";
        }
    }
}
} // namespace

int main() {
    std::vector<std::pair<int, int>> max_by_delta(1000000);
    for (int target_height = 1; target_height <= 999899; ++target_height) {
        Analyse(target_height, max_by_delta);
    }
    std::cout << "\n";
    int farmax = 0;
    for (int delta = 1; delta <= 999999; ++delta) {
        farmax = std::max(farmax, max_by_delta[delta].first);
        std::cout << "delta=" << delta << ": " << (delta + max_by_delta[delta].second) << "->" << max_by_delta[delta].second << ": " << max_by_delta[delta].first << " steps (max " << farmax << ")\n";
    }
}
```

And found:

```
delta=1: 2->1: 1 steps (max 1)
delta=2: 4->2: 2 steps (max 2)
delta=3: 5->2: 3 steps (max 3)
delta=4: 9->5: 4 steps (max 4)
delta=6: 8->2: 5 steps (max 5)
delta=7: 9->2: 6 steps (max 6)
delta=9: 11->2: 7 steps (max 7)
delta=14: 16->2: 8 steps (max 8)
delta=15: 17->2: 9 steps (max 9)
delta=17: 19->2: 10 steps (max 10)
delta=25: 35->10: 11 steps (max 11)
delta=30: 32->2: 12 steps (max 12)
delta=31: 33->2: 13 steps (max 13)
delta=33: 35->2: 14 steps (max 14)
delta=42: 44->2: 15 steps (max 15)
delta=58: 60->2: 16 steps (max 16)
delta=63: 65->2: 17 steps (max 17)
delta=65: 67->2: 18 steps (max 18)
delta=74: 76->2: 19 steps (max 19)
delta=90: 92->2: 20 steps (max 20)
delta=122: 156->34: 21 steps (max 21)
delta=127: 129->2: 22 steps (max 22)
delta=129: 131->2: 23 steps (max 23)
delta=138: 140->2: 24 steps (max 24)
delta=154: 156->2: 25 steps (max 25)
delta=250: 252->2: 26 steps (max 26)
delta=255: 257->2: 27 steps (max 27)
delta=257: 259->2: 28 steps (max 28)
delta=266: 268->2: 29 steps (max 29)
delta=282: 284->2: 30 steps (max 30)
delta=378: 380->2: 31 steps (max 31)
delta=506: 540->34: 32 steps (max 32)
delta=511: 513->2: 33 steps (max 33)
delta=513: 515->2: 34 steps (max 34)
delta=522: 524->2: 35 steps (max 35)
delta=538: 540->2: 36 steps (max 36)
delta=634: 636->2: 37 steps (max 37)
delta=1014: 1016->2: 38 steps (max 38)
delta=1023: 1025->2: 39 steps (max 39)
delta=1025: 1027->2: 40 steps (max 40)
delta=1034: 1036->2: 41 steps (max 41)
delta=1050: 1052->2: 42 steps (max 42)
delta=1146: 1148->2: 43 steps (max 43)
delta=1526: 1528->2: 44 steps (max 44)
delta=2038: 2040->2: 45 steps (max 45)
delta=2047: 2049->2: 46 steps (max 46)
delta=2049: 2051->2: 47 steps (max 47)
delta=2058: 2060->2: 48 steps (max 48)
delta=2074: 2076->2: 49 steps (max 49)
delta=2170: 2172->2: 50 steps (max 50)
delta=2550: 2552->2: 51 steps (max 51)
delta=3062: 3064->2: 52 steps (max 52)
delta=4086: 4088->2: 53 steps (max 53)
delta=4097: 4099->2: 54 steps (max 54)
delta=4106: 4108->2: 55 steps (max 55)
delta=4122: 4124->2: 56 steps (max 56)
delta=4218: 4220->2: 57 steps (max 57)
delta=4598: 4600->2: 58 steps (max 58)
delta=5110: 5112->2: 59 steps (max 59)
delta=6134: 6136->2: 60 steps (max 60)
delta=8182: 8184->2: 61 steps (max 61)
delta=8193: 8195->2: 62 steps (max 62)
delta=8202: 8204->2: 63 steps (max 63)
delta=8218: 8220->2: 64 steps (max 64)
delta=8314: 8316->2: 65 steps (max 65)
delta=8694: 8696->2: 66 steps (max 66)
delta=9206: 9208->2: 67 steps (max 67)
delta=10230: 10232->2: 68 steps (max 68)
delta=12278: 12280->2: 69 steps (max 69)
delta=16374: 16376->2: 70 steps (max 70)
delta=16394: 16396->2: 71 steps (max 71)
delta=16410: 16412->2: 72 steps (max 72)
delta=16506: 16508->2: 73 steps (max 73)
delta=16886: 16888->2: 74 steps (max 74)
delta=17398: 17400->2: 75 steps (max 75)
delta=18422: 18424->2: 76 steps (max 76)
delta=20470: 20472->2: 77 steps (max 77)
delta=24566: 24568->2: 78 steps (max 78)
delta=32758: 32760->2: 79 steps (max 79)
delta=32778: 32780->2: 80 steps (max 80)
delta=32794: 32796->2: 81 steps (max 81)
delta=32890: 32892->2: 82 steps (max 82)
delta=33270: 33272->2: 83 steps (max 83)
delta=33782: 33784->2: 84 steps (max 84)
delta=34806: 34808->2: 85 steps (max 85)
delta=36854: 36856->2: 86 steps (max 86)
delta=40950: 40952->2: 87 steps (max 87)
delta=49142: 49144->2: 88 steps (max 88)
delta=65526: 65528->2: 89 steps (max 89)
delta=65562: 65564->2: 90 steps (max 90)
delta=65658: 65660->2: 91 steps (max 91)
delta=66038: 66040->2: 92 steps (max 92)
delta=66550: 66552->2: 93 steps (max 93)
delta=67574: 67576->2: 94 steps (max 94)
delta=69622: 69624->2: 95 steps (max 95)
delta=73718: 73720->2: 96 steps (max 96)
delta=81910: 81912->2: 97 steps (max 97)
delta=98294: 98296->2: 98 steps (max 98)
delta=131062: 131064->2: 99 steps (max 99)
delta=131098: 131100->2: 100 steps (max 100)
delta=131194: 131196->2: 101 steps (max 101)
delta=131574: 131576->2: 102 steps (max 102)
delta=132086: 132088->2: 103 steps (max 103)
delta=133110: 133112->2: 104 steps (max 104)
delta=135158: 135160->2: 105 steps (max 105)
delta=139254: 139256->2: 106 steps (max 106)
delta=147446: 147448->2: 107 steps (max 107)
delta=163830: 163832->2: 108 steps (max 108)
delta=196598: 196600->2: 109 steps (max 109)
delta=262134: 262136->2: 110 steps (max 110)
delta=262266: 262268->2: 111 steps (max 111)
delta=262646: 262648->2: 112 steps (max 112)
delta=263158: 263160->2: 113 steps (max 113)
delta=264182: 264184->2: 114 steps (max 114)
delta=266230: 266232->2: 115 steps (max 115)
delta=270326: 270328->2: 116 steps (max 116)
delta=278518: 278520->2: 117 steps (max 117)
delta=294902: 294904->2: 118 steps (max 118)
delta=327670: 327672->2: 119 steps (max 119)
delta=393206: 393208->2: 120 steps (max 120)
delta=524278: 524280->2: 121 steps (max 121)
delta=524410: 524412->2: 122 steps (max 122)
delta=524790: 524792->2: 123 steps (max 123)
delta=525302: 525304->2: 124 steps (max 124)
delta=526326: 526328->2: 125 steps (max 125)
delta=528374: 528376->2: 126 steps (max 126)
delta=532470: 532472->2: 127 steps (max 127)
delta=540662: 540664->2: 128 steps (max 128)
delta=557046: 557048->2: 129 steps (max 129)
delta=589814: 589816->2: 130 steps (max 130)
delta=655350: 655352->2: 131 steps (max 131)
delta=786422: 786424->2: 132 steps (max 132)
```

l0rinc commented:
ACK 0f6b58ba9a2989feac6292e74e2d721e35492051

Since last ack the formula was adjusted to be more precise.
This documents the change better, it provides more accurate documentation. Thanks for persevering!

sedited reviewed (commented):
Concept ACK

The first commit seems like a good thing to exercise, but I'm not sure about the latter two. Do we really need to validate this internal behaviour?

sipa commented:
After more consideration, I think this PR is too wide in scope, and some of the things it attempts to do aren't really the right place for it. I apologize for having contributed to the expansion.

I think there are at least 4 distinct properties relevant here:
1. Does `GetSkipHeight` work correctly?
2. Do the `CBlockIndex::pskip` pointers get set correctly (according to `GetSkipHeight`).
3. Does the skip-pointer-following algorithm have good complexity characteristics?
4. Does `GetAncestor` implement the skip-pointer-following algorithm correctly?

* The first commit tests a combination of (2) + (3), however, I think 3 is more research-level question that doesn't need a unit test. It could be a document, explanation, perhaps even Python code for analysing alternatives, as it's unrelated to the actual code.
* The third commit tests (1), but that absolutely trivial and exposing the function just for this seems overkill, and possibly even a performance reduction.
* Nothing tests (4), though it would be the most interesting thing to test. However, it's probably too high a cost (it'd need some exposing the algorithm in a way that allows injecting tracing logic).

What I would suggest, if you want to pursue this, is to just test (1) + (2), which can be do done without any exposing: create a test chain, and verify that the heights the pskip pointers at specific heights are expected values (e.g. the values in your current third commit).

optout21 commented:
Thanks for this analysis, it clarifies things!
Based on the previous discussions, I agree that a unit test is not best suited for testing the complexity of the GetSkipHeight/GetAncestor (4) algorithm (for one, it doesn't check the actual code, but a copy of it). I also agree that the simple unit test (1) is trivial, and the need to expose an internal function doesn't worth it (I've mentioned this doubt previously).
I will rework the test to check the actual skip values by checking the `pskip` pointers, as you suggested, and skip the algorithmic complexity test (number-of-steps-needed).

optout21 committed:
test: Add test for BuildSkip() and skip heights

Add new test that calls `CBlockIndex::BuildSkip()` and verifies skip height values.
It tests that:
- `pprev` field is set (to an earlier block index),
- the skip is to the index as dictated by the `GetSkipHeight()` bit-manipulation logic.

Co-authored-by: sipa <pieter@wuille.net>
Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

optout21 commented:
Reworked, following https://github.com/bitcoin/bitcoin/pull/33661#issuecomment-4479391151.
I have de-scoped the test for the upper-bound of the steps needed by `GetAncestor()`, which is essentially an algorithmic complexity test, and a unit test is not suitable for it.
The test on the skip height logic has been moved to a higher level: the same skip values are tested, but as retrieved from the set `pskip` pointers, set by `BuildSkip()`.

There is only one commit now (test-only).

The old commits:

- 1st was the algorithm complexity test, dropped
- 2nd was exposing `GetSkipHeight()`, no longer needed, dropped
- 3rd: the trivial unit test is dropped, replaced by the new test.

The PR description has been updated.
I was considering opening a new PR, but I decided to keep the current one.
Used a recent baseline.
Original commits archived [here](https://github.com/optout21/bitcoin/commits/skip-height-test3/).

sipa commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,103 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        block_index[i].BuildSkip();
+    }
+
+    for (const std::pair<int, int>& tp : TEST_DATA) {
```
Simpler:

```c++
for (auto [input, expected] : TEST_DATA) {
```

(`TEST_DATA` can even be inlined here as just `{{0b00010010, 0b00010000 }, {0b00100010, 0b00100000}, ...}`, but feel free to keep it as a separate variable above)

optout21 replied:
Done

sipa commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,103 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+    const std::vector<std::pair<int, int>> TEST_DATA = {
+        // EVEN values: the rightmost set bit is zeroed
+        // Various even values with at least 2 bits set
+        { 0b00010010 ,
```
Ultra-nitty style comment: Splitting each test cases over two (super short) lines seems excessive to me.

l0rinc replied:
The point is to diff them easily - otherwise the binary format itself doesn't make sense

optout21 replied:
The rationale is that the changed bits -- skip calculation involves zeroing 1 or 2 bits -- are more visible by aligning the result beneath the input. That's why binary format is used (and auto-formatting is turned off). This was requested by @l0rinc, and it makes sense.

sipa reviewed (commented):
Thank you for understanding and the quick turnaround.

ACK 47b7c675b69b7407c36170b8619ed098a0f41da5. Only two small nits.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,103 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        BOOST_REQUIRE_LT(expected, input);
+
+        BOOST_CHECK(block_index[input].pskip);
+        const int skip_height = block_index[input].pskip->nHeight;
```
nit: we're using the safer brace init elsewhere:
```suggestion
        const int skip_height{block_index[input].pskip->nHeight};
```

optout21 replied:
Done.

optout21 reviewed (commented):
Minor change applied following review comment.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+    // - `GetAncestor()` works as expected (indirectly).
+
+    // Build a chain (up to the highest test input value)
+    constexpr int chain_size{27'482};
```
nit: alternatively, to avoid the magic number and the code comment:

```suggestion
    const auto max_test_input{std::ranges::max_element(TEST_DATA, [](auto& a, auto& b) { return a.first < b.first; })};
    const int chain_size{max_test_input->first + 1};
```

optout21 replied:
Applied. I was expecting a comment for such an improvement :)
Note that this requires `<algorithm>`.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        BOOST_REQUIRE_GT(input, 0);
+        BOOST_REQUIRE_LT(expected, input);
+
+        BOOST_CHECK(block_index[input].pskip);
```
nit: technically this seems more like a require, i.e. a precondition:
```suggestion
        BOOST_REQUIRE(block_index[input].pskip);
```

optout21 replied:
The `pskip` pointer is set by the `BuildSkip()` call, which is the object of the test; that's why I went with the CHECK.
It's a bit misleading that the `BuildSkip()` call is in the first loop with the preparations. It can't be in the second loop, because it has to be called on all the elements.
To make this more clear, I moved this `BOOST_CHECK` check to right after the `BuildSkip()` call (with the extra qualification for `i==0`, as there `pskip` is `null`).

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        { 19,  1 },
+        { 20, 16 },
+    };
+    // clang-format on
```
nit: we don't usually need this since there is no automatic formatting

optout21 replied:
OK, I see, but I prefer to keep it, it shows the intent at least.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+    }
+
+    for (auto [input, expected] : TEST_DATA) {
+        // test data sanity check
```
nit: the require already signals "sanity check", I personally don't find the comment adds anything

optout21 replied:
Dropped.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+BOOST_AUTO_TEST_CASE(build_skip_height_test)
+{
+    // clang-format off
+    const std::vector<std::pair<int, int>> TEST_DATA = {
```
This is fixed size, we should be able to make it an array instead:

```suggestion
    constexpr std::array<std::pair<int, int>, 36> TEST_DATA{{
```

nit: the mixed equals and brace init seems a bit weird to me
nit2: these may need some new includes

optout21 replied:
While an array is simpler, I don't like the need for the explicit count. However, a plain C++ `[]` array also works and that doesn't need the explicit count, so I changed to use that. (It also doesn't need any include.)

l0rinc replied:
Note: [`std::array`](https://cppreference.com/cpp/container/array) is the C++ fixed-size container equivalent of this C-style array, and avoids array-to-pointer decay when passed around. That said, this one stays local and only used as a range here, so the C array seems fine in this instance.

optout21 replied:
Of course, `std::array` has many advantages, but I think in this test-only local-only usage an old-style array is fine.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +188,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        block_index[i].BuildSkip();
+    }
+
+    for (auto [input, expected] : TEST_DATA) {
```
nit: can probably be reference:

```suggestion
    for (auto& [input, expected] : TEST_DATA) {
```

optout21 replied:
Done.

l0rinc reviewed (approved):
ACK e5aeda01cdd906f3989847cc56de5de2af5dc0c3

Left a few nits, happy to re-review if you decide to apply them.
I'm also fine with merging as-is. Thanks for your persistence!

optout21 commented:
Minor changes applied, following the detailed review -- thanks @l0rinc !

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +189,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+BOOST_AUTO_TEST_CASE(build_skip_height_test)
+{
+    // clang-format off
+    const std::pair<int, int> TEST_DATA[]{
```
nit: `std::pair` requires `#include <utility>`, see https://en.cppreference.com/cpp/utility/pair

optout21 replied:
Added.

l0rinc commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +189,100 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+        block_index[i].pprev = i == 0 ? nullptr : &block_index[i - 1];
+        block_index[i].nHeight = i;
+        block_index[i].BuildSkip();
+        BOOST_CHECK(block_index[i].pskip || i == 0);
```
nit: this should likely be `BOOST_REQUIRE` instead, otherwise if that invariant ever fails, the test may continue into a `nullptr` dereference instead of reporting a clean Boost failure

optout21 replied:
Fair; I've added back the `BOOST_REQUIRE` for the `pskip` pointer before accessing it.

l0rinc reviewed (approved):
ACK e64fb9ec94f8f297d8288195774121e7791d3b4b

optout21 reviewed (commented):
Applied changes: two more nits (after new review from @l0rinc, some response comments to previous review showed up again, apologies.).

l0rinc commented:
ACK 131fa570b9551c1bfa6a4844b15dc8ee3e443201

Since last ack, `pskip` is required to avoid possible `nullptr` dereference and added missing include + previous pending comments submitted.

sipa commented on src/test/skiplist_tests.cpp:
```diff
@@ -188,4 +190,101 @@ BOOST_AUTO_TEST_CASE(findearliestatleast_edge_test)
+    // - `GetAncestor()` works as expected (indirectly).
+
+    // Build a chain (up to the highest test input value)
+    const auto max_test_input{std::ranges::max_element(TEST_DATA, [](auto& a, auto& b) { return a.first < b.first; })};
```
Nit: could be `static constexpr auto MAX_TEST_INPUT = std::ranges::max_element(TEST_DATA)`, if `TEST_DATA` is also made `static constexpr`.

sipa reviewed (commented):
ACK 131fa570b9551c1bfa6a4844b15dc8ee3e443201

l0rinc commented:
rfm?
